File: blue-sdk-android/blue-sdk/src/main/kotlin/com/blue/sdk/internal/CommandQueue.kt
// CommandQueue.kt
// BlueSDK - 指令串行队列（ARCH-04）
// 同一时刻只允许一条指令在等待应答，支持多指令 FIFO 排队
// 超时 5 秒自动重试，最多重试 3 次（NFR02）

package com.blue.sdk.internal

import android.os.Handler
import android.os.Looper
import com.blue.sdk.error.BlueError
import com.blue.sdk.transport.CommandCode
import com.blue.sdk.transport.FrameConstants
import com.blue.sdk.transport.ParsedFrame
import java.util.LinkedList

internal class CommandQueue {

    companion object {
        const val DEFAULT_TIMEOUT_MS = 5000L
        const val DEFAULT_MAX_RETRIES = 3
    }

    private data class PendingCommand(
        val cmd: Byte,
        val frame: ByteArray,
        val retryCount: Int,
        val maxRetries: Int,
        val timeoutMs: Long,
        val completion: (Result<ParsedFrame>) -> Unit
    )

    private var pendingCommand: PendingCommand? = null   // 当前等待应答的指令
    private val waitingQueue = LinkedList<PendingCommand>() // FIFO 等待队列
    private val lock = Any()
    private val handler = Handler(Looper.getMainLooper())
    private var timeoutRunnable: Runnable? = null

    /** 发送指令的实际执行块（由 BLEConnector 注入）*/
    var sendBlock: ((ByteArray) -> Unit)? = null

    fun enqueue(
        cmd: Byte,
        frame: ByteArray,
        timeoutMs: Long = DEFAULT_TIMEOUT_MS,
        maxRetries: Int = DEFAULT_MAX_RETRIES,
        completion: (Result<ParsedFrame>) -> Unit
    ) {
        synchronized(lock) {
            val command = PendingCommand(cmd, frame, 0, maxRetries, timeoutMs, completion)
            if (pendingCommand == null) {
                pendingCommand = command
                send(command)
            } else {
                waitingQueue.add(command)
            }
        }
    }

    /**
     * 直接发送帧，不经过队列排队和等待应答（用于时间同步等 fire-and-forget 场景）
     */
    fun sendDirect(frame: ByteArray) {
        BlueLogger.debug("直接发送: ${frame.size} 字节")
        sendBlock?.invoke(frame)
    }

    fun handleResponse(frame: ParsedFrame): Boolean {
        synchronized(lock) {
            val pending = pendingCommand ?: return false
            val cmdInt = pending.cmd.toInt() and 0xFF
            val respInt = frame.cmd.toInt() and 0xFF
            val isMatch = respInt == cmdInt + 1 || respInt == cmdInt
            if (!isMatch) return false
            
            // 对于 sendCommand(0x06) 的应答(0x07)，需匹配 DPID 避免主动上报帧误匹配
            if (cmdInt == (CommandCode.SEND_COMMAND.toInt() and 0xFF) &&
                respInt == (CommandCode.DEVICE_REPORT.toInt() and 0xFF)) {
                val pendingDPID = if (pending.frame.size > FrameConstants.DATA_OFFSET)
                    pending.frame[FrameConstants.DATA_OFFSET] else null
                val responseDPID = frame.data.firstOrNull()
                if (pendingDPID != null && responseDPID != null && pendingDPID != responseDPID) {
                    return false
                }
            }

            cancelTimeout()
            pendingCommand = null
            pending.completion(Result.success(frame))
            sendNext()
            return true
        }
    }

    fun clear() {
        synchronized(lock) {
            cancelTimeout()
            pendingCommand?.let { it.completion(Result.failure(BlueError.Disconnected)) }
            pendingCommand = null
            waitingQueue.forEach { it.completion(Result.failure(BlueError.Disconnected)) }
            waitingQueue.clear()
        }
    }

    private fun send(command: PendingCommand) {
        BlueLogger.debug("发送指令: 0x${"%02X".format(command.cmd.toInt() and 0xFF)} (重试${command.retryCount})")
        val block = sendBlock
        if (block == null) {
            // 没有连接，立即回调 Disconnected 错误
            pendingCommand = null
            command.completion(Result.failure(BlueError.Disconnected))
            sendNext()
            return
        }
        block.invoke(command.frame)
        scheduleTimeout(command)
    }

    private fun sendNext() {
        if (waitingQueue.isEmpty()) return
        val next = waitingQueue.poll() ?: return
        pendingCommand = next
        // 指令间隔至少 200ms
        handler.postDelayed({ send(next) }, 200)
    }

    private fun scheduleTimeout(command: PendingCommand) {
        cancelTimeout()
        val runnable = Runnable { handleTimeout() }
        timeoutRunnable = runnable
        handler.postDelayed(runnable, command.timeoutMs)
    }

    private fun handleTimeout() {
        synchronized(lock) {
            val command = pendingCommand ?: return
            if (command.retryCount < command.maxRetries) {
                val retried = command.copy(retryCount = command.retryCount + 1)
                pendingCommand = retried
                send(retried)
            } else {
                pendingCommand = null
                command.completion(Result.failure(BlueError.Timeout))
                sendNext()
            }
        }
    }

    private fun cancelTimeout() {
        timeoutRunnable?.let { handler.removeCallbacks(it) }
        timeoutRunnable = null
    }
}

File: blue-sdk-ios/BlueSDK/BlueSDK/Classes/Internal/CommandQueue.swift
// CommandQueue.swift
// BlueSDK - LX-PD02 智能药盒蓝牙通信 SDK
//
// 指令串行队列：同一时刻只允许一条指令在等待应答（ARCH-04）
// 超时 5 秒自动重试，最多重试 3 次（NFR02）
// 支持多指令排队，按 FIFO 顺序依次发送

import Foundation

/// 待执行指令
private struct PendingCommand {
    let cmd: UInt8
    let frame: [UInt8]
    let retryCount: Int
    let maxRetries: Int
    let timeout: TimeInterval
    let completion: (Result<ParsedFrame, BlueError>) -> Void
}

/// 指令串行队列
/// 管理下行指令的发送、超时重试和应答匹配
final class CommandQueue {

    // MARK: - 配置

    static let defaultTimeout: TimeInterval = 5.0
    static let defaultMaxRetries: Int = 3

    // MARK: - 状态

    private var pendingCommand: PendingCommand?   // 当前正在等待应答的指令
    private var waitingQueue: [PendingCommand] = [] // 等待发送的指令队列
    private var timeoutTimer: Timer?
    private let lock = NSLock()

    /// 发送指令的实际执行块（由 BLEConnector 注入）
    var sendBlock: (([UInt8]) -> Void)?

    /// 直接发送帧数据，不入队等待应答（用于单向下发指令如时间同步）
    func sendDirect(frame: [UInt8]) {
        sendBlock?(frame)
    }

    // MARK: - 入队

    /// 将指令加入队列
    /// - Parameters:
    ///   - cmd: 命令字
    ///   - frame: 完整帧数据
    ///   - timeout: 超时时间，默认 5 秒
    ///   - maxRetries: 最大重试次数，默认 3 次
    ///   - completion: 完成回调
    func enqueue(
        cmd: UInt8,
        frame: [UInt8],
        timeout: TimeInterval = defaultTimeout,
        maxRetries: Int = defaultMaxRetries,
        completion: @escaping (Result<ParsedFrame, BlueError>) -> Void
    ) {
        lock.lock()
        defer { lock.unlock() }

        let command = PendingCommand(
            cmd: cmd,
            frame: frame,
            retryCount: 0,
            maxRetries: maxRetries,
            timeout: timeout,
            completion: completion
        )

        if pendingCommand == nil {
            // 队列空闲，直接发送
            pendingCommand = command
            send(command)
        } else {
            // 有指令在等待，加入等待队列
            waitingQueue.append(command)
        }
    }

    // MARK: - 应答处理

    /// 处理设备应答帧
    @discardableResult
    func handleResponse(_ frame: ParsedFrame) -> Bool {
        lock.lock()

        guard let pending = pendingCommand else {
            lock.unlock()
            return false
        }

        // CMD 匹配：应答帧 CMD == 发送帧 CMD + 1 或相同
        let cmdMatch = frame.cmd == pending.cmd + 1 || frame.cmd == pending.cmd
        guard cmdMatch else {
            lock.unlock()
            return false
        }

        // 对于 sendCommand(0x06) 的应答(0x07)，还需匹配 DPID（数据第一字节）
        // 避免设备主动上报帧（如低电 0x75）被错误匹配
        if pending.cmd == CommandCode.sendCommand, frame.cmd == CommandCode.deviceReport {
            let pendingDPID = pending.frame.count > Int(FrameConstants.dataOffset)
                ? pending.frame[Int(FrameConstants.dataOffset)]
                : nil
            let responseDPID = frame.data.first
            if let pDPID = pendingDPID, let rDPID = responseDPID, pDPID != rDPID {
                lock.unlock()
                return false
            }
        }

        cancelTimeout()
        pendingCommand = nil
        // 先释放锁，再调用 completion（防止 completion 内再次 enqueue 导致死锁）
        lock.unlock()

        pending.completion(.success(frame))

        // 发送队列中的下一条指令
        lock.lock()
        sendNext()
        lock.unlock()
        return true
    }

    /// 清空队列，取消所有待处理指令
    func clear() {
        lock.lock()
        defer { lock.unlock() }
        cancelTimeout()
        pendingCommand?.completion(.failure(.disconnected))
        pendingCommand = nil
        waitingQueue.forEach { $0.completion(.failure(.disconnected)) }
        waitingQueue.removeAll()
    }

    // MARK: - 私有方法

    private func send(_ command: PendingCommand) {
        guard let block = sendBlock else {
            // 没有连接，立即回调 Disconnected 错误
            pendingCommand = nil
            command.completion(.failure(.disconnected))
            sendNext()
            return
        }
        block(command.frame)
        scheduleTimeout(for: command)
    }

    private func sendNext() {
        guard !waitingQueue.isEmpty else { return }
        let next = waitingQueue.removeFirst()
        pendingCommand = next
        // 指令间隔至少 200ms，避免设备来不及处理
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) { [weak self] in
            self?.send(next)
        }
    }

    private func scheduleTimeout(for command: PendingCommand) {
        cancelTimeout()
        timeoutTimer = Timer.scheduledTimer(
            withTimeInterval: command.timeout,
            repeats: false
        ) { [weak self] _ in
            self?.handleTimeout()
        }
    }

    private func handleTimeout() {
        lock.lock()
        guard let command = pendingCommand else {
            lock.unlock()
            return
        }

        if command.retryCount < command.maxRetries {
            let retried = PendingCommand(
                cmd: command.cmd,
                frame: command.frame,
                retryCount: command.retryCount + 1,
                maxRetries: command.maxRetries,
                timeout: command.timeout,
                completion: command.completion
            )
            pendingCommand = retried
            lock.unlock()
            send(retried)
        } else {
            pendingCommand = nil
            lock.unlock()
            command.completion(.failure(.timeout))
            // 继续发送队列中的下一条
            lock.lock()
            sendNext()
            lock.unlock()
        }
    }

    private func cancelTimeout() {
        timeoutTimer?.invalidate()
        timeoutTimer = nil
    }
}
